### PR TITLE
[ir] Add CHI Builder code examples

### PR DESCRIPTION
**Pull request by BillXu2000**
Related issue = #

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

Add tests as CHI Builder code examples, including struct for, external array and autograd.

### CODE REVIEW

**Review by k-ye — DISMISSED** [review]
<s>Great LGTM!</s>

On a second thought, this isn't really a test. I was hoping to get a standalone executable in some `demos/`. This serves as a minimum demonstration for how users can use Taichi's CPP library in their own.

File: tests/cpp/ir/ir_builder_test.cpp [hunk 1]
```diff
@@ -7,6 +7,287 @@
...
+ext_arr = np.zeros(n, np.int32)
+ext(ext_arr)
+#ext_arr = place.to_numpy()
+print(ext_arr)
+*/
+
+  auto program = Program(arch_from_name("x64"));
+  /*CompileConfig config_print_ir;
+  config_print_ir.print_ir = true;
+  prog_.config = config_print_ir;*/  // print_ir = True
```
Comment (xumingkuan) [inline tests/cpp/ir/ir_builder_test.cpp:48]: (IIUC)
```suggestion
  program.config = config_print_ir;*/  // print_ir = True
```

File: tests/cpp/ir/ir_builder_test.cpp [hunk 2]
```diff
@@ -7,6 +7,287 @@
 namespace taichi {
 namespace lang {
 
+TEST(IRBuilder, RunSnode) {
+/*
+import taichi as ti, numpy as np
+ti.init()
+#ti.init(print_ir = True)
```
Comment (xumingkuan) [inline tests/cpp/ir/ir_builder_test.cpp:14]: nit:
```suggestion
# ti.init(print_ir=True)
```

File: tests/cpp/ir/ir_builder_test.cpp [hunk 3]
```diff
@@ -7,6 +7,287 @@
...
+
+n = 10
+place = ti.field(dtype = ti.i32)
```
Comment (xumingkuan) [inline tests/cpp/ir/ir_builder_test.cpp:17]: nit:
```suggestion
place = ti.field(dtype=ti.i32)
```

File: tests/cpp/ir/ir_builder_test.cpp [hunk 4]
```diff
@@ -7,6 +7,287 @@
...
+@ti.kernel
+def ext(ext_arr: ti.ext_arr()):
+    for index in place:
+        ext_arr[index] = place[index]
+
+init()
+print(ret())
+ext_arr = np.zeros(n, np.int32)
+ext(ext_arr)
+#ext_arr = place.to_numpy()
```
Comment (xumingkuan) [inline tests/cpp/ir/ir_builder_test.cpp:41]: nit:
```suggestion
# ext_arr = place.to_numpy()
```

File: tests/cpp/ir/ir_builder_test.cpp [hunk 5]
```diff
@@ -7,6 +7,287 @@
...
+TEST(IRBuilder, Autograd) {
+/*
+import taichi as ti, numpy as np
+ti.init()
+
+n = 10
+a = ti.field(ti.f32, n, needs_grad = True)
+b = ti.field(ti.f32, n, needs_grad = True)
+c = ti.field(ti.f32, n, needs_grad = True)
+energy = ti.field(ti.f32, [], needs_grad = True)
```
Comment (xumingkuan) [inline tests/cpp/ir/ir_builder_test.cpp:148]: nit:
```suggestion
a = ti.field(ti.f32, n, needs_grad=True)
b = ti.field(ti.f32, n, needs_grad=True)
c = ti.field(ti.f32, n, needs_grad=True)
energy = ti.field(ti.f32, [], needs_grad=True)
```

**Review by xumingkuan — COMMENTED** [review]
I agree that this is not a test. But maybe it's easier to write a `TEST(...)` since it doesn't need to change the .cmake files?

**Comment by k-ye** [comment]
> I agree that this is not a test. But maybe it's easier to write a `TEST(...)` since it doesn't need to change the .cmake files?

The purpose is to demo how taichi's `so` can be included into other person's existing CPP project. In that regard it is better to have a standalone binary.

**Comment by BillXu2000** [comment]
\format

File: examples/chi_examples/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,16 @@
+cmake_minimum_required(VERSION 3.12)
+
+set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DTI_ISE_NONE -std=c++17 -fsized-deallocation -Wall  -march=nehalem -DTI_ARCH_x64 -DTI_PASS_EXCEPTION_TO_PYTHON -DTI_INCLUDED -DTI_WITH_CUDA -DTI_WITH_OPENGL -DTI_WITH_CC -O3 -DNDEBUG -static-libgcc -static-libstdc++ -Wl,--wrap=log2f -rdynamic")
```
Comment (k-ye) [inline examples/chi_examples/CMakeLists.txt:3]:
```suggestion
set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++17 -fsized-deallocation -Wall  -march=nehalem -DTI_ARCH_x64 -DTI_INCLUDED -O3 -DNDEBUG")
```

Could you try this?

**Review by k-ye — APPROVED** [review]

File: examples/chi_examples/README.md [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+Follow the steps in `https://taichi.readthedocs.io/en/stable/dev_install.html`.
+
+Add option `-DTI_EXPORT_CORE=ON` to your `cmake` command (i.e. use `cmake .. DTI_EXPORT_CORE=ON`).
+
+Make sure taichi is built under `$TAICHI_REPO_DIR/build` directory.
+
+After building, `$TAICHI_REPO_DIR/build/libtaichi_export_core.so` should exist.
+
+## Link with Taichi Shared Library
```
Comment (xumingkuan) [inline examples/chi_examples/README.md:13]:
```suggestion
## Link with the Taichi Shared Library
```

File: cmake/TaichiExportCore.cmake [hunk 8]
```diff
@@ -0,0 +1,6 @@
+cmake_minimum_required(VERSION 3.0)
```
Comment (xumingkuan) [inline cmake/TaichiExportCore.cmake:1]: Why does this require v3.0 while examples/chi_examples/CMakeLists.txt requires v3.12?

**Review by xumingkuan — APPROVED** [review]
LGTM!

**Pull request merged**